streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -469,48 +465,76 @@ private void pipeRecord(final String topicName,
         }
         final TopicPartition topicPartition = getTopicPartition(topicName);
-        if (topicPartition != null) {
-            final long offset = offsetsByTopicPartition.get(topicPartition).incrementAndGet() - 1;
-            task.addRecords(topicPartition, Collections.singleton(new ConsumerRecord<>(
-                    topicName,
-                    topicPartition.partition(),
-                    offset,
-                    timestamp,
-                    TimestampType.CREATE_TIME,
-                    (long) ConsumerRecord.NULL_CHECKSUM,
-                    key == null ? ConsumerRecord.NULL_SIZE : key.length,
-                    value == null ? ConsumerRecord.NULL_SIZE : value.length,
-                    key,
-                    value,
-                    headers)));
-
-            // Process the record ...
-            task.process(mockWallClockTime.milliseconds());
-            task.maybePunctuateStreamTime();
-            task.commit();
-            captureOutputRecords();
+
+        if (topicPartition == null) {
+            processGlobalRecord(topicName, timestamp, key, value, headers);
         } else {
-            final TopicPartition globalTopicPartition = globalPartitionsByTopic.get(topicName);
-            if (globalTopicPartition == null) {
-                throw new IllegalArgumentException("Unknown topic: " + topicName);
+            enqueueTaskRecord(topicName, topicPartition, timestamp, key, value, headers);
+            processAllProcessableRecords();
+        }
+    }
+
+    private void enqueueTaskRecord(final String topicName,
+                                   final TopicPartition topicPartition,
+                                   final Long timestamp,
+                                   final byte[] key,
+                                   final byte[] value,
+                                   final Headers headers) {
+        if (!internalTopologyBuilder.sourceTopicNames().isEmpty()) {
+            validateSourceTopicNameRegexPattern(topicName);
+        }
+        Objects.requireNonNull(topicPartition);
+
+        final long offset = offsetsByTopicPartition.get(topicPartition).incrementAndGet() - 1;
+        task.addRecords(topicPartition, Collections.singleton(new ConsumerRecord<>(
+            topicName,
+            topicPartition.partition(),
+            offset,
+            timestamp,
+            TimestampType.CREATE_TIME,
+            (long) ConsumerRecord.NULL_CHECKSUM,
+            key == null ? ConsumerRecord.NULL_SIZE : key.length,
+            value == null ? ConsumerRecord.NULL_SIZE : value.length,
+            key,
+            value,
+            headers)));
+    }
+
+    private void processAllProcessableRecords() {
+        // If the topology only has global tasks, then `task` would be null.
+        // For this method, it just means there's nothing to do.
+        if (task != null) {
+            while (task.hasRecordsQueued()) {
```

Comment:
There is already `StreamTask#numBuffered()` -- should we reuse this one instead of adding a new method?

Reply:
That one is private. I considered it, but it seemed silly to expose `integer numBuffered()` when the only thing we'd do with it is `while (numBuffered() > 0)`. If we have another reason to expose the integer in the future, we can merge the use cases later on.